File: branches/OpenRemote_Boss_2_0_DeveloperReleases/iPhone_Console_2_0_0_Alpha1/Classes/ButtonView.m
/* OpenRemote, the Home of the Digital Home.
 * Copyright 2008-2009, OpenRemote Inc.
 * 
 * See the contributors.txt file in the distribution for a
 * full listing of individual contributors.
 * 
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3.0 of
 * the License, or (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

#import "ButtonView.h"
#import "Control.h"
#import "DirectoryDefinition.h"
#import "ServerDefinition.h"
#import "ViewHelper.h"
#import "NotificationConstant.h"
#import "ClippedUIImage.h"

//defines the interval (seconds) of command when pressing a repeat button
#define REPEAT_CMD_INTERVAL 0.3

@interface ButtonView (Private) 
- (void)createButton;
- (void)controlButtonUp:(id)sender;
- (void)controlButtonDown:(id)sender;
- (void)sendCommand:(id)sender;

@end

@implementation ButtonView

@synthesize uiButton, uiImage, uiImagePressed;

#pragma mark Private methods

//Create button according to control and add tap event
- (void)createButton {
	if (uiButton != nil) {
		[uiButton release];
		uiButton = nil;
	}
	uiButton = [[UIButton buttonWithType:UIButtonTypeCustom] retain];
	
	[uiButton addTarget:self action:@selector(controlButtonDown:) forControlEvents:UIControlEventTouchDown];	
	//[uiButton addTarget:self action:@selector(controlButtonUp:) forControlEvents:UIControlEventTouchUpOutside];	
	[uiButton addTarget:self action:@selector(controlButtonUp:) forControlEvents:UIControlEventTouchUpInside];

	[self addSubview:uiButton];
	
}

- (void) controlButtonUp:(id)sender {
	[self cancelTimer];
	Button *button = (Button *)component;
	if (button.navigate) {
		[[NSNotificationCenter defaultCenter] postNotificationName:NotificationNavigateTo object:button.navigate];
	}
}

- (void) controlButtonDown:(id)sender {
	[self cancelTimer];
	
	Button *button = (Button *)component;
	if (button.hasCommand == YES) {
		[self sendCommand:nil];
	 	if (button.repeat == YES ) {			
			controlTimer = [NSTimer scheduledTimerWithTimeInterval:REPEAT_CMD_INTERVAL	target:self selector:@selector(sendCommand:) userInfo:nil repeats:YES];			
		} 
	}

}

- (void) sendCommand:(id)sender {
	[self	sendCommandRequest:@"click"];
}

#pragma mark Override the methods of superclass(ComponentView)

- (void)initView {	
	[self createButton];
	
	Button *button = (Button *)component;
	if (button.defaultImage) {
		uiImage = [[UIImage alloc] initWithContentsOfFile:[[DirectoryDefinition imageCacheFolder] stringByAppendingPathComponent:button.defaultImage.src]];
		uiImagePressed = [[UIImage alloc] initWithContentsOfFile:[[DirectoryDefinition imageCacheFolder] stringByAppendingPathComponent:button.pressedImage.src]];	
		ClippedUIImage *clippedUIImage = [[ClippedUIImage alloc] initWithUIImage:uiImage dependingOnUIView:self imageAlignToView:IMAGE_ABSOLUTE_ALIGN_TO_VIEW];		
		[uiButton setImage:clippedUIImage forState:UIControlStateNormal];
		if (uiImagePressed) {
			ClippedUIImage *clippedUIImagePressed = [[ClippedUIImage alloc] initWithUIImage:uiImagePressed dependingOnUIView:self imageAlignToView:IMAGE_ABSOLUTE_ALIGN_TO_VIEW];
			[uiButton setImage:clippedUIImagePressed forState:UIControlStateHighlighted];
		}
		//use top-left alignment
		[uiButton setFrame:CGRectMake(0, 0, clippedUIImage.size.width, clippedUIImage.size.height)];
	} else {
		[uiButton setFrame:[self bounds]];
		UIImage *buttonImage = [[UIImage imageNamed:@"button.png"] stretchableImageWithLeftCapWidth:20 topCapHeight:20];
		[uiButton setBackgroundImage:buttonImage forState:UIControlStateNormal];
		
		//buttonImage = [[UIImage imageNamed:@"buttonHighlighted.png"] stretchableImageWithLeftCapWidth:20 topCapHeight:20];
		//[uiButton setBackgroundImage:buttonImage forState:UIControlStateHighlighted];
		
		uiButton.titleLabel.font = [UIFont boldSystemFontOfSize:18];
		[uiButton setTitleShadowColor:[UIColor grayColor] forState:UIControlStateNormal];
		uiButton.titleLabel.shadowOffset = CGSizeMake(0, -2);
		[uiButton setTitle:button.name forState:UIControlStateNormal];
	}
	
}

#pragma mark dealloc

- (void)dealloc {
	[uiImage  release];
	[uiImagePressed release];
	[uiButton release];
	
  [super dealloc];
}


@end

File: workspace/richturner/Android_Console_2_0_0_Betas/src/org/openremote/android/console/view/ButtonView.java
/* OpenRemote, the Home of the Digital Home.
* Copyright 2008-2011, OpenRemote Inc.
*
* See the contributors.txt file in the distribution for a
* full listing of individual contributors.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Affero General Public License as
* published by the Free Software Foundation, either version 3 of the
* License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU Affero General Public License for more details.
*
* You should have received a copy of the GNU Affero General Public License
* along with this program. If not, see <http://www.gnu.org/licenses/>.
*/

package org.openremote.android.console.view;

import java.util.Timer;
import java.util.TimerTask;

import org.openremote.android.console.Constants;
import org.openremote.android.console.R;
import org.openremote.android.console.bindings.ORButton;
import org.openremote.android.console.model.ListenerConstant;
import org.openremote.android.console.model.ORListenerManager;
import org.openremote.android.console.util.ImageUtil;

import android.content.Context;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.BitmapDrawable;
import android.text.TextUtils.TruncateAt;
import android.util.TypedValue;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.FrameLayout;

/**
 * 
 * 
 * @author <a href="mailto:richard@openremote.org">Richard Turner</a>
 * @author ?
 */
public class ButtonView extends ControlView {

   private Button uiButton;
   private BitmapDrawable defaultImage;
   private BitmapDrawable pressedImage;
   
   /** The Constant REPEAT_CMD_INTERVAL. */
   public final static long REPEAT_CMD_INTERVAL = 300;
   
   public ButtonView(Context context, ORButton button) {
      super(context);
      setComponent(button);
      if (button != null) {
         uiButton  = new Button(context, null, android.R.attr.buttonStyleSmall);
         initButton(button);
      }
   }
   
   /**
    * Inits the button.
    * Add listener to make the button send command and navigate to.
    * If pressed, send command; if up, navigate to. It also can 
    * repeat send command if pressed and not up.  
    * 
    * @param button the button
    */
   private void initButton(final ORButton button) {
      int width = button.getFrameWidth();
      int height = button.getFrameHeight();
      uiButton.setEllipsize(TruncateAt.END);
      uiButton.setLines(1);
      uiButton.setId(button.getComponentId());
      uiButton.setText(button.getName());
      uiButton.setPadding(0, 0, 0, 0);
      uiButton.setTypeface(Typeface.SANS_SERIF, Typeface.BOLD);
      uiButton.setTextSize(TypedValue.COMPLEX_UNIT_DIP, Constants.BUTTON_FONT_SIZE_DIP);
      uiButton.setBackgroundDrawable(getResources().getDrawable(R.drawable.button));
      FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(width, height);
      uiButton.setLayoutParams(params);

      if (button.getDefaultImage() != null) {
         defaultImage = ImageUtil.createClipedDrawableFromPath(context, Constants.FILE_FOLDER_PATH + button.getDefaultImage().getSrc(), width, height);
         if (defaultImage != null) {
            uiButton.setBackgroundDrawable(defaultImage);
         }
      }
      
      if (button.getPressedImage() != null) {
         pressedImage = ImageUtil.createClipedDrawableFromPath(context, Constants.FILE_FOLDER_PATH + button.getPressedImage().getSrc(), width, height);
      }
      View.OnTouchListener touchListener = new OnTouchListener() {
         public boolean onTouch(View v, MotionEvent event) {
            if (event.getAction() == MotionEvent.ACTION_DOWN) {
               cancelTimer();
               if (pressedImage != null) {
                  uiButton.setBackgroundDrawable(pressedImage);
               } else if (defaultImage != null) {
                  defaultImage.setAlpha(200);
                  uiButton.setBackgroundDrawable(defaultImage);
               }
               if (button.isHasControlCommand()) {
                  sendCommand();
                  if (button.isRepeat()) {
                     Timer timer = new Timer();
                     timer.schedule(new TimerTask() {
                        public void run() {
                           sendCommand();
                        }
                     }, REPEAT_CMD_INTERVAL, REPEAT_CMD_INTERVAL);
                     setTimer(timer);
                  }
               }
            } else if (event.getAction() == MotionEvent.ACTION_UP) {
               cancelTimer();
               if (defaultImage != null) {
                  defaultImage.setAlpha(255);
                  uiButton.setBackgroundDrawable(defaultImage);
               }
               if (button.getNavigate() != null) {
                  uiButton.setPressed(false);
                  ORListenerManager.getInstance().notifyOREventListener(ListenerConstant.ListenerNavigateTo, button.getNavigate());
               }
            }
            return false;
         }
      };
      uiButton.setOnTouchListener(touchListener);
      addView(uiButton);
   }

   private void sendCommand() {
      sendCommandRequest("click");
   }
}
